### PR TITLE
VSTS 438 - User Notifications - Frontend

### PR DESCRIPTION
**Pull request by jessedoyle**
:elephant: :art:

* Add a `AmaLayout::NotificationDecorator` to handle view-specific logic for individual notifications.
* Add an engine-level controller action that allows any application to dismiss all notifications via a simple API call (`DELETE /api/v1/notifications`).
* Add a coffeescript `AMALayout.Notifications` class that is responsible for calling the notification dismiss API when certain user actions are triggered in the DOM.
* Remove the `Rails` namespace from `AmaLayout::Rails::Engine` - it was redundant.
* Add a `NotificationScrubber` that will properly and safely strip HTML nodes from the notification content before rendering it.
* Add a basic `require_login` and `current_user` stub to the internal test application - we need to test behaviour that is behind a login page.
* Add various partials for notifications and an off canvas notification center when logged in.
* Add full test coverage for new behaviour.
* Add a `delete` method to the `NotificationSet`. We need to be able to remove notifications fully when they are no longer applicable (i.e. user has already renewed).
* Add a `brand` attribute to notifications. We need an easy way to select batches of similar notifications (i.e. when deleting). We could use the `SHA256` digest to do so, but then we'd have to populate all of the content of a notification. Selecting on `brand` will be simpler for most use cases. This is an optional attribute an is nilable.
* Make the `lifespan` attribute not a factor into the `SHA256` digest of a notification. The default value of `1.year` in seconds changes between Rails 4 (31557600) and Rails 5.1 (31556952). We don't want the version of Rails to affect the SHA digest at all.

#### Unread Notifications in Nav Bar

screen shot 2017-06-30 at 2 02 03 pm

#### Notification Center

screen shot 2017-06-30 at 2 02 12 pm

### CODE REVIEW

**Comment by MathieuGilbert** [comment]
📖

File: lib/ama_layout/notification_set.rb [hunk 1]
```diff
@@ -20,7 +21,7 @@ class NotificationSet
     include Enumerable
     attr_accessor :base, :data_store, :key
 
-    delegate :each, :first, :last, :size, :[], :empty?, :any?,  to: :active
+    delegate :each, :first, :last, :size, :[], :empty?, :any?,  to: :all
```
Comment (MathieuGilbert) [inline lib/ama_layout/notification_set.rb]: Is this delegation still needed if they're acting on `all`?  I might be missing something but it seems like that should be the default.

File: spec/ama_layout/notification_scrubber_spec.rb [hunk 2]
```diff
@@ -0,0 +1,10 @@
+describe AmaLayout::NotificationScrubber do
+  describe '#initialize' do
+    let(:sanitized) { Loofah.fragment(string).scrub!(subject).to_s }
+    let(:string) { '<script>alert("haxxed");</script><a href="#" invalid="test">test</a>waffles' }
+
+    it 'scrubs HTML tags from a string' do
+      expect(sanitized).to eq('alert("haxxed");<a href="#">test</a>waffles')
```
Comment (MathieuGilbert) [inline spec/ama_layout/notification_scrubber_spec.rb]: 👍

File: spec/ama_layout/decorators/navigation_decorator_spec.rb [hunk 3]
```diff
@@ -129,4 +138,53 @@
       expect(navigation_presenter.account_toggle).to eq "render"
     end
   end
+
+  context 'notification center' do
+    let(:any?) { true }
+    let(:user) { OpenStruct.new(navigation: 'member', notifications: notifications) }
+    let(:notifications) { OpenStruct.new(active: [notification], all: [notification], any?: any?) }
```
Comment (MathieuGilbert) [inline spec/ama_layout/decorators/navigation_decorator_spec.rb]: Could you also cover some "0" and "2" count cases? If the `active` and `all` arrays are variables, you won't have to override `any?` either.

**Comment by MathieuGilbert** [comment]
💩

File: lib/ama_layout/decorators/navigation_decorator.rb [hunk 4]
```diff
@@ -32,5 +32,56 @@ def name_or_email
...
+      end
+    end
+
+    private
+
+    def decorated_notifications
+      AmaLayout::NotificationDecorator.decorate_collection(user.notifications)
+    end
+
+    def notifications?
```
Comment (jessedoyle) [inline lib/ama_layout/decorators/navigation_decorator.rb]: Yes - that's more simple.

File: lib/ama_layout/decorators/navigation_decorator.rb [hunk 5]
```diff
@@ -32,5 +32,56 @@ def name_or_email
...
+      user.present?
+    end
+
+    def notification_count
+      notifications? && user.notifications.active.size || 0
+    end
+
+    def badge?
```
Comment (jessedoyle) [inline lib/ama_layout/decorators/navigation_decorator.rb]: Sure!

File: app/views/ama_layout/_notification_sidebar.html.erb [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+    </button>
+  </div>
+  <div class="row column">
+    <%= navigation.notifications_heading %>
+  </div>
+  <%= render partial: 'ama_layout/notification', collection: notifications %>
+  <div class="row column">
+    <p class="mt1">Did You Know?</p>
+    <div class="section-content">
+      <p class="bold">Renew Your Membership Automatically</p>
```
Comment (jessedoyle) [inline app/views/ama_layout/_notification_sidebar.html.erb]: I'm working with the PO/business unit and it's looking like this is going to switch to a VRAR promotion instead.

**Comment by jessedoyle** [comment]
:wave:

File: lib/ama_layout/notification_set.rb (hunk 1 above)
Comment (jessedoyle) [inline lib/ama_layout/notification_set.rb]: We spoke about this in person. As of right now the delegation is needed as many tests are written assuming a `NotificationSet` responds to array-like methods.

File: app/views/ama_layout/_notification_sidebar.html.erb [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+  <div class="row column">
+    <%= navigation.notifications_heading %>
+  </div>
+  <%= render partial: 'ama_layout/notification', collection: notifications %>
+  <div class="row column">
+    <p class="mt1">Did You Know?</p>
+    <div class="section-content">
+      <p class="bold">Sign up for Vehicle Registration Auto-Renew</p>
+      <p>
+        We’ll process your payment and mail you a new registration certificate and licence plate sticker. It’s that simple!
```
Comment (MathieuGilbert) [inline app/views/ama_layout/_notification_sidebar.html.erb]: Could you replace the `’`s with `'`?

**Comment by MathieuGilbert** [comment]
💩

File: app/views/ama_layout/_notification_sidebar.html.erb (hunk 7 above)
Comment (jessedoyle) [inline app/views/ama_layout/_notification_sidebar.html.erb]: Good call! Those always sneak through!

**Comment by jessedoyle** [comment]
:wave:

**Comment by MathieuGilbert** [comment]
:elephant: :elephant: :elephant:

**Comment by MathieuGilbert** [comment]
:elephant: :elephant: :elephant:

**Comment by kaytcampbell** [comment]
:art: :art: :art:

**Pull request merged**